File: Android/app/src/androidTest/java/jp/co/riso/smartdeviceapp/model/printsettings/PreviewTest.java

package jp.co.riso.smartdeviceapp.model.printsettings;

import android.test.AndroidTestCase;
import android.util.Log;

import org.w3c.dom.Document;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

import java.io.IOException;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.List;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import jp.co.riso.android.util.AppUtils;
import jp.co.riso.smartdeviceapp.AppConstants;
import jp.co.riso.smartdeviceapp.SmartDeviceApp;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.BookletFinish;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.BookletLayout;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.ColorMode;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.Duplex;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.FinishingSide;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.Imposition;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.ImpositionOrder;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.Orientation;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.OutputTray;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.PaperSize;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.Punch;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.Sort;
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.Staple;

public class PreviewTest extends AndroidTestCase {
    private static final String TAG = "PreviewTest";
    private NodeList mSettingList;

    @Override
    protected void setUp() throws Exception {
        super.setUp();
        String xmlString = AppUtils.getFileContentsFromAssets(SmartDeviceApp.getAppContext(),
                AppConstants.XML_FILENAME);

        Document printSettingsContent = null;

        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
        try {
            DocumentBuilder db = dbf.newDocumentBuilder();

            InputSource is = new InputSource();
            is.setCharacterStream(new StringReader(xmlString));
            printSettingsContent = db.parse(is);
        } catch (ParserConfigurationException e) {
            Log.e(TAG, "Error: " + e.getMessage());
        } catch (SAXException e) {
            Log.e(TAG, "Error: " + e.getMessage());
        } catch (IOException e) {
            Log.e(TAG, "Error: " + e.getMessage());
        }

        if (printSettingsContent == null) {
            return;
        }

        mSettingList = printSettingsContent.getElementsByTagName("setting");
    }

    @Override
    protected void tearDown() throws Exception {
        super.tearDown();
    }

    public void testConstructor() {
        assertNotNull(new Preview());
    }

    public void testColorMode() {
        assertEquals(0, ColorMode.AUTO.ordinal());
        assertEquals(1, ColorMode.FULL_COLOR.ordinal());
        assertEquals(2, ColorMode.MONOCHROME.ordinal());
        assertEquals(3, ColorMode.DUAL_COLOR.ordinal());
        assertEquals(ColorMode.AUTO, ColorMode.valueOf("AUTO"));
        assertEquals(ColorMode.FULL_COLOR, ColorMode.valueOf("FULL_COLOR"));
        assertEquals(ColorMode.MONOCHROME, ColorMode.valueOf("MONOCHROME"));
    }

    public void testOrientation() {
        assertEquals(0, Orientation.PORTRAIT.ordinal());
        assertEquals(1, Orientation.LANDSCAPE.ordinal());
        assertEquals(Orientation.PORTRAIT, Orientation.valueOf("PORTRAIT"));
        assertEquals(Orientation.LANDSCAPE, Orientation.valueOf("LANDSCAPE"));
    }

    public void testDuplex() {
        assertEquals(0, Duplex.OFF.ordinal());
        assertEquals(1, Duplex.LONG_EDGE.ordinal());
        assertEquals(2, Duplex.SHORT_EDGE.ordinal());
        assertEquals(Duplex.OFF, Duplex.valueOf("OFF"));
        assertEquals(Duplex.LONG_EDGE, Duplex.valueOf("LONG_EDGE"));
        assertEquals(Duplex.SHORT_EDGE, Duplex.valueOf("SHORT_EDGE"));
    }

    public void testPaperSizeValueOf() {
        assertEquals(PaperSize.A3, PaperSize.valueOf("A3"));
    }

    public void testPaperSizeGetWidth() {
        assertEquals(297.0f, PaperSize.A3.getWidth());
        assertEquals(210.0f, PaperSize.A4.getWidth());
    }

    public void testPaperSizeGetHeight() {
        assertEquals(420.0f, PaperSize.A3.getHeight());
        assertEquals(297.0f, PaperSize.A4.getHeight());
    }

    public void testImpositionValueOf() {
        assertEquals(Imposition.OFF, Imposition.valueOf("OFF"));
    }

    public void testImpositionGetPerPage() {
        assertEquals(1, Imposition.OFF.getPerPage());
    }

    public void testImpositionGetRows() {
        assertEquals(1, Imposition.OFF.getRows());
    }

    public void testImpositionGetCols() {
        assertEquals(1, Imposition.OFF.getCols());
    }

    public void testImpositionIsFlipLandscape() {
        assertEquals(false, Imposition.OFF.isFlipLandscape());
    }

    public void testImpositionOrderIsLeftToRight() {
        assertEquals(true, ImpositionOrder.L_R.isLeftToRight());
    }

    public void testImpositionOrderIsTopToBottom() {
        assertEquals(true, ImpositionOrder.L_R.isTopToBottom());
    }

    public void testImpositionOrderIsHorizontalFlow() {
        assertEquals(true, ImpositionOrder.L_R.isHorizontalFlow());
    }

    public void testImpositionOrderValueOf() {
        assertEquals(ImpositionOrder.L_R, ImpositionOrder.valueOf("L_R"));
    }

    public void testSort() {
        assertEquals(0, Sort.PER_PAGE.ordinal());
        assertEquals(1, Sort.PER_COPY.ordinal());
        assertEquals(Sort.PER_PAGE, Sort.valueOf("PER_PAGE"));
        assertEquals(Sort.PER_COPY, Sort.valueOf("PER_COPY"));
    }

    public void testBookletFinish() {
        assertEquals(0, BookletFinish.OFF.ordinal());
        assertEquals(1, BookletFinish.PAPER_FOLDING.ordinal());
        assertEquals(2, BookletFinish.FOLD_AND_STAPLE.ordinal());
        assertEquals(BookletFinish.OFF, BookletFinish.valueOf("OFF"));
        assertEquals(BookletFinish.PAPER_FOLDING, BookletFinish.valueOf("PAPER_FOLDING"));
        assertEquals(BookletFinish.FOLD_AND_STAPLE, BookletFinish.valueOf("FOLD_AND_STAPLE"));
    }

    public void testBookletLayout() {
        assertEquals(0, BookletLayout.FORWARD.ordinal());
        assertEquals(1, BookletLayout.REVERSE.ordinal());
        assertEquals(BookletLayout.FORWARD, BookletLayout.valueOf("FORWARD"));
        assertEquals(BookletLayout.REVERSE, BookletLayout.valueOf("REVERSE"));
    }

    public void testFinishingSide() {
        assertEquals(0, FinishingSide.LEFT.ordinal());
        assertEquals(1, FinishingSide.TOP.ordinal());
        assertEquals(2, FinishingSide.RIGHT.ordinal());
        assertEquals(FinishingSide.LEFT, FinishingSide.valueOf("LEFT"));
        assertEquals(FinishingSide.TOP, FinishingSide.valueOf("TOP"));
        assertEquals(FinishingSide.RIGHT, FinishingSide.valueOf("RIGHT"));
    }

    public void testStapleValueOf() {
        assertEquals(Staple.OFF, Staple.valueOf("OFF"));
    }

    public void testStapleGetCount() {
        assertEquals(0, Staple.OFF.getCount());
        assertEquals(1, Staple.ONE_UL.getCount());
        assertEquals(2, Staple.TWO.getCount());
    }

    public void testPunchValueOf() {
        assertEquals(Punch.OFF, Punch.valueOf("OFF"));
    }

    public void testPunchGetCount() {
        assertEquals(0, Punch.OFF.getCount(false));
        assertEquals(2, Punch.HOLES_2.getCount(false));
        assertEquals(4, Punch.HOLES_4.getCount(false));
    }

    public void testPunchGetCount_JapanesePrinter() {
        assertEquals(0, Punch.OFF.getCount(true));
        assertEquals(2, Punch.HOLES_2.getCount(true));
        assertEquals(3, Punch.HOLES_4.getCount(true));
    }

    public void testOutputTray() {
        assertEquals(0, OutputTray.AUTO.ordinal());
        assertEquals(1, OutputTray.FACEDOWN.ordinal());
        assertEquals(2, OutputTray.TOP.ordinal());
        assertEquals(3, OutputTray.STACKING.ordinal());
        assertEquals(OutputTray.AUTO, OutputTray.valueOf("AUTO"));
        assertEquals(OutputTray.FACEDOWN, OutputTray.valueOf("FACEDOWN"));
        assertEquals(OutputTray.TOP, OutputTray.valueOf("TOP"));
        assertEquals(OutputTray.STACKING, OutputTray.valueOf("STACKING"));
    }

    public void testColorMode_XML() {
        NodeList optionsNodeList  = mSettingList.item(0).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();
        
        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), ColorMode.values().length - 1); // Default does not have dual color mode

    }

    public void testOrientation_XML() {
        NodeList optionsNodeList  = mSettingList.item(1).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();

        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), Orientation.values().length);
    }

    public void testDuplex_XML() {
        NodeList optionsNodeList  = mSettingList.item(3).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();

        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), Duplex.values().length);
    }

    public void testPaperSize_XML() {
        NodeList optionsNodeList  = mSettingList.item(4).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();

        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), PaperSize.values().length-3); // Only tests IS
    }

    public void testImposition_XML() {
        NodeList optionsNodeList  = mSettingList.item(8).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();

        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), Imposition.values().length);
    }

    public void testImpositionOrder_XML() {
        NodeList optionsNodeList  = mSettingList.item(9).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();

        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), ImpositionOrder.values().length);
    }

    public void testSort_XML() {
        NodeList optionsNodeList  = mSettingList.item(10).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();

        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), Sort.values().length);
    }

    public void testBookletFinish_XML() {
        NodeList optionsNodeList  = mSettingList.item(12).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();

        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), BookletFinish.values().length);
    }

    public void testBookletLayout_XML() {
        NodeList optionsNodeList  = mSettingList.item(13).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();
;
        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), BookletLayout.values().length);
    }

    public void testFinishingSide_XML() {
        NodeList optionsNodeList  = mSettingList.item(14).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();

        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), FinishingSide.values().length);
    }

    public void testStaple_XML() {
        NodeList optionsNodeList  = mSettingList.item(15).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();

        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), Staple.values().length);
    }

    public void testPunch_XML() {
        NodeList optionsNodeList  = mSettingList.item(16).getChildNodes();
        List<Option> optionsList = new ArrayList<Option>();

        for (int i = 1; i < optionsNodeList.getLength(); i += 2) {
            optionsList.add(new Option(optionsNodeList.item(i)));
        }
        assertEquals(optionsList.size(), Punch.values().length);
    }

}

File: Android/app/src/androidTest/java/jp/co/riso/smartdeviceapp/model/printsettings/PreviewTest.kt
package jp.co.riso.smartdeviceapp.model.printsettings

import android.util.Log
import jp.co.riso.android.util.AppUtils
import jp.co.riso.smartdeviceapp.AppConstants
import jp.co.riso.smartdeviceapp.SmartDeviceApp
import jp.co.riso.smartdeviceapp.model.printsettings.Preview.*
import junit.framework.TestCase
import org.junit.Before
import org.junit.Test
import org.w3c.dom.Document
import org.w3c.dom.NodeList
import org.xml.sax.InputSource
import org.xml.sax.SAXException
import java.io.IOException
import java.io.StringReader
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException

class PreviewTest {

    private var _settingList: NodeList? = null

    @Before
    fun setUp() {
        val xmlString = AppUtils.getFileContentsFromAssets(
            SmartDeviceApp.getAppContext(),
            AppConstants.XML_FILENAME
        )
        var printSettingsContent: Document? = null
        val dbf = DocumentBuilderFactory.newInstance()
        try {
            val db = dbf.newDocumentBuilder()
            val `is` = InputSource()
            `is`.characterStream = StringReader(xmlString)
            printSettingsContent = db.parse(`is`)
        } catch (e: ParserConfigurationException) {
            Log.e(TAG, "Error: " + e.message)
        } catch (e: IOException) {
            Log.e(TAG, "Error: " + e.message)
        } catch (e: SAXException) {
            Log.e(TAG, "Error: " + e.message)
        }
        if (printSettingsContent == null) {
            return
        }
        _settingList = printSettingsContent.getElementsByTagName("setting")
    }

    @Test
    fun testConstructor() {
        TestCase.assertNotNull(Preview())
    }

    @Test
    fun testColorMode() {
        TestCase.assertEquals(0, ColorMode.AUTO.ordinal)
        TestCase.assertEquals(1, ColorMode.FULL_COLOR.ordinal)
        TestCase.assertEquals(2, ColorMode.MONOCHROME.ordinal)
        TestCase.assertEquals(3, ColorMode.DUAL_COLOR.ordinal)
        TestCase.assertEquals(ColorMode.AUTO, ColorMode.valueOf("AUTO"))
        TestCase.assertEquals(ColorMode.FULL_COLOR, ColorMode.valueOf("FULL_COLOR"))
        TestCase.assertEquals(ColorMode.MONOCHROME, ColorMode.valueOf("MONOCHROME"))
    }

    @Test
    fun testOrientation() {
        TestCase.assertEquals(0, Orientation.PORTRAIT.ordinal)
        TestCase.assertEquals(1, Orientation.LANDSCAPE.ordinal)
        TestCase.assertEquals(Orientation.PORTRAIT, Orientation.valueOf("PORTRAIT"))
        TestCase.assertEquals(
            Orientation.LANDSCAPE,
            Orientation.valueOf("LANDSCAPE")
        )
    }

    @Test
    fun testDuplex() {
        TestCase.assertEquals(0, Duplex.OFF.ordinal)
        TestCase.assertEquals(1, Duplex.LONG_EDGE.ordinal)
        TestCase.assertEquals(2, Duplex.SHORT_EDGE.ordinal)
        TestCase.assertEquals(Duplex.OFF, Duplex.valueOf("OFF"))
        TestCase.assertEquals(Duplex.LONG_EDGE, Duplex.valueOf("LONG_EDGE"))
        TestCase.assertEquals(Duplex.SHORT_EDGE, Duplex.valueOf("SHORT_EDGE"))
    }

    @Test
    fun testPaperSizeValueOf() {
        TestCase.assertEquals(PaperSize.A3, PaperSize.valueOf("A3"))
    }

    @Test
    fun testPaperSizeGetWidth() {
        TestCase.assertEquals(297.0f, PaperSize.A3.width)
        TestCase.assertEquals(210.0f, PaperSize.A4.width)
    }

    @Test
    fun testPaperSizeGetHeight() {
        TestCase.assertEquals(420.0f, PaperSize.A3.height)
        TestCase.assertEquals(297.0f, PaperSize.A4.height)
    }

    @Test
    fun testInputTray() {
        TestCase.assertEquals(0, InputTray_FT_GL_CEREZONA_S.AUTO.ordinal)
        TestCase.assertEquals(1, InputTray_FT_GL_CEREZONA_S.STANDARD.ordinal)
        TestCase.assertEquals(2, InputTray_FT_GL_CEREZONA_S.TRAY1.ordinal)
        TestCase.assertEquals(3, InputTray_FT_GL_CEREZONA_S.TRAY2.ordinal)
        TestCase.assertEquals(4, InputTray_FT_GL_CEREZONA_S.TRAY3.ordinal)
        TestCase.assertEquals(5, InputTray_FT_GL_CEREZONA_S.EXTERNAL_FEEDER.ordinal)
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.AUTO,
            InputTray_FT_GL_CEREZONA_S.valueOf("AUTO")
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.STANDARD,
            InputTray_FT_GL_CEREZONA_S.valueOf("STANDARD")
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.EXTERNAL_FEEDER,
            InputTray_FT_GL_CEREZONA_S.valueOf("EXTERNAL_FEEDER")
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.AUTO,
            InputTray_FT_GL_CEREZONA_S.valuesFT_CEREZONA_S()[0]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.STANDARD,
            InputTray_FT_GL_CEREZONA_S.valuesFT_CEREZONA_S()[1]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.TRAY1,
            InputTray_FT_GL_CEREZONA_S.valuesFT_CEREZONA_S()[2]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.TRAY2,
            InputTray_FT_GL_CEREZONA_S.valuesFT_CEREZONA_S()[3]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.EXTERNAL_FEEDER,
            InputTray_FT_GL_CEREZONA_S.valuesFT_CEREZONA_S()[4]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.AUTO,
            InputTray_FT_GL_CEREZONA_S.valuesGL()[0]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.STANDARD,
            InputTray_FT_GL_CEREZONA_S.valuesGL()[1]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.TRAY1,
            InputTray_FT_GL_CEREZONA_S.valuesGL()[2]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.TRAY2,
            InputTray_FT_GL_CEREZONA_S.valuesGL()[3]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.TRAY3,
            InputTray_FT_GL_CEREZONA_S.valuesGL()[4]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.EXTERNAL_FEEDER,
            InputTray_FT_GL_CEREZONA_S.valuesGL()[5]
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.AUTO,
            InputTray_FT_GL_CEREZONA_S.valueOf("AUTO")
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.STANDARD,
            InputTray_FT_GL_CEREZONA_S.valueOf("STANDARD")
        )
        TestCase.assertEquals(
            InputTray_FT_GL_CEREZONA_S.EXTERNAL_FEEDER,
            InputTray_FT_GL_CEREZONA_S.valueOf("EXTERNAL_FEEDER")
        )
    }

    @Test
    fun testImpositionValueOf() {
        TestCase.assertEquals(Imposition.OFF, Imposition.valueOf("OFF"))
    }

    @Test
    fun testImpositionGetPerPage() {
        TestCase.assertEquals(1, Imposition.OFF.perPage)
    }

    @Test
    fun testImpositionGetRows() {
        TestCase.assertEquals(1, Imposition.OFF.rows)
    }

    @Test
    fun testImpositionGetCols() {
        TestCase.assertEquals(1, Imposition.OFF.cols)
    }

    @Test
    fun testImpositionIsFlipLandscape() {
        TestCase.assertEquals(false, Imposition.OFF.isFlipLandscape)
    }

    @Test
    fun testImpositionOrderIsLeftToRight() {
        TestCase.assertEquals(true, ImpositionOrder.L_R.isLeftToRight)
    }

    @Test
    fun testImpositionOrderIsTopToBottom() {
        TestCase.assertEquals(true, ImpositionOrder.L_R.isTopToBottom)
    }

    @Test
    fun testImpositionOrderIsHorizontalFlow() {
        TestCase.assertEquals(true, ImpositionOrder.L_R.isHorizontalFlow)
    }

    @Test
    fun testImpositionOrderValueOf() {
        TestCase.assertEquals(ImpositionOrder.L_R, ImpositionOrder.valueOf("L_R"))
    }

    @Test
    fun testSort() {
        TestCase.assertEquals(0, Sort.PER_COPY.ordinal)
        TestCase.assertEquals(1, Sort.PER_PAGE.ordinal)
        TestCase.assertEquals(Sort.PER_PAGE, Sort.valueOf("PER_PAGE"))
        TestCase.assertEquals(Sort.PER_COPY, Sort.valueOf("PER_COPY"))
    }

    @Test
    fun testBookletFinish() {
        TestCase.assertEquals(0, BookletFinish.OFF.ordinal)
        TestCase.assertEquals(1, BookletFinish.PAPER_FOLDING.ordinal)
        TestCase.assertEquals(2, BookletFinish.FOLD_AND_STAPLE.ordinal)
        TestCase.assertEquals(BookletFinish.OFF, BookletFinish.valueOf("OFF"))
        TestCase.assertEquals(BookletFinish.PAPER_FOLDING, BookletFinish.valueOf("PAPER_FOLDING"))
        TestCase.assertEquals(
            BookletFinish.FOLD_AND_STAPLE,
            BookletFinish.valueOf("FOLD_AND_STAPLE")
        )
    }

    @Test
    fun testBookletLayout() {
        TestCase.assertEquals(0, BookletLayout.FORWARD.ordinal)
        TestCase.assertEquals(1, BookletLayout.REVERSE.ordinal)
        TestCase.assertEquals(BookletLayout.FORWARD, BookletLayout.valueOf("FORWARD"))
        TestCase.assertEquals(BookletLayout.REVERSE, BookletLayout.valueOf("REVERSE"))
    }

    @Test
    fun testFinishingSide() {
        TestCase.assertEquals(0, FinishingSide.LEFT.ordinal)
        TestCase.assertEquals(1, FinishingSide.TOP.ordinal)
        TestCase.assertEquals(2, FinishingSide.RIGHT.ordinal)
        TestCase.assertEquals(FinishingSide.LEFT, FinishingSide.valueOf("LEFT"))
        TestCase.assertEquals(FinishingSide.TOP, FinishingSide.valueOf("TOP"))
        TestCase.assertEquals(FinishingSide.RIGHT, FinishingSide.valueOf("RIGHT"))
    }

    @Test
    fun testStapleValueOf() {
        TestCase.assertEquals(Staple.OFF, Staple.valueOf("OFF"))
    }

    @Test
    fun testStapleGetCount() {
        TestCase.assertEquals(0, Staple.OFF.count)
        TestCase.assertEquals(1, Staple.ONE_UL.count)
        TestCase.assertEquals(2, Staple.TWO.count)
    }

    @Test
    fun testPunchValueOf() {
        TestCase.assertEquals(Punch.OFF, Punch.valueOf("OFF"))
    }

    @Test
    fun testPunchGetCount() {
        TestCase.assertEquals(0, Punch.OFF.getCount(false))
        TestCase.assertEquals(2, Punch.HOLES_2.getCount(false))
        TestCase.assertEquals(4, Punch.HOLES_4.getCount(false))
    }

    @Test
    fun testPunchGetCount_JapanesePrinter() {
        TestCase.assertEquals(0, Punch.OFF.getCount(true))
        TestCase.assertEquals(2, Punch.HOLES_2.getCount(true))
        TestCase.assertEquals(3, Punch.HOLES_4.getCount(true))
    }

    @Test
    fun testOutputTray() {
        TestCase.assertEquals(0, OutputTray.AUTO.ordinal)
        TestCase.assertEquals(1, OutputTray.FACEDOWN.ordinal)
        TestCase.assertEquals(2, OutputTray.TOP.ordinal)
        TestCase.assertEquals(3, OutputTray.STACKING.ordinal)
        TestCase.assertEquals(OutputTray.AUTO, OutputTray.valueOf("AUTO"))
        TestCase.assertEquals(OutputTray.FACEDOWN, OutputTray.valueOf("FACEDOWN"))
        TestCase.assertEquals(OutputTray.TOP, OutputTray.valueOf("TOP"))
        TestCase.assertEquals(OutputTray.STACKING, OutputTray.valueOf("STACKING"))
    }

    @Test
    fun testColorMode_XML() {
        val optionsNodeList = _settingList!!.item(0).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(
            optionsList.size,
            ColorMode.values().size - 1
        ) // Default does not have dual color mode
    }

    @Test
    fun testOrientation_XML() {
        val optionsNodeList = _settingList!!.item(1).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(optionsList.size, Orientation.values().size)
    }

    @Test
    fun testDuplex_XML() {
        val optionsNodeList = _settingList!!.item(3).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(optionsList.size, Duplex.values().size)
    }

    @Test
    fun testPaperSize_XML() {
        val optionsNodeList = _settingList!!.item(4).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(
            optionsList.size,
            PaperSize.values().size - 4
        ) // Only tests IS
    }

    @Test
    fun testImposition_XML() {
        val optionsNodeList = _settingList!!.item(8).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(optionsList.size, Imposition.values().size)
    }

    @Test
    fun testImpositionOrder_XML() {
        val optionsNodeList = _settingList!!.item(9).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(optionsList.size, ImpositionOrder.values().size)
    }

    @Test
    fun testSort_XML() {
        val optionsNodeList = _settingList!!.item(10).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(optionsList.size, Sort.values().size)
    }

    @Test
    fun testBookletFinish_XML() {
        val optionsNodeList = _settingList!!.item(12).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(optionsList.size, BookletFinish.values().size)
    }

    @Test
    fun testBookletLayout_XML() {
        val optionsNodeList = _settingList!!.item(13).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(optionsList.size, BookletLayout.values().size)
    }

    @Test
    fun testFinishingSide_XML() {
        val optionsNodeList = _settingList!!.item(14).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(optionsList.size, FinishingSide.values().size)
    }

    @Test
    fun testStaple_XML() {
        val optionsNodeList = _settingList!!.item(15).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(optionsList.size, Staple.values().size)
    }

    @Test
    fun testPunch_XML() {
        val optionsNodeList = _settingList!!.item(16).childNodes
        val optionsList: MutableList<Option> = ArrayList()
        var i = 1
        while (i < optionsNodeList.length) {
            optionsList.add(Option(optionsNodeList.item(i)))
            i += 2
        }
        TestCase.assertEquals(optionsList.size, Punch.values().size)
    }

    companion object {
        private const val TAG = "PreviewTest"
    }
}